channels/channels.go

```diff
@@ -362,29 +392,74 @@ func (c *Channels) HasChannel(chid datatransfer.ChannelID) (bool, error) {
 // - a corresponding "progress" event if the block has not been seen before
+// - a pause event if the progress goes past the data limit
 // For example, if a block is being sent for the first time, the method will
 // fire both DataSent AND DataSentProgress.
 // If a block is resent, the method will fire DataSent but not DataSentProgress.
-// Returns true if the block is new (both the event and a progress event were fired).
-func (c *Channels) fireProgressEvent(chid datatransfer.ChannelID, evt datatransfer.EventCode, progressEvt datatransfer.EventCode, k cid.Cid, delta uint64, index int64, unique bool, readFromOriginal readOriginalFn) (bool, error) {
+// If a block is sent for the first time, and more data has been sent than the data limit,
+// the method will fire DataSent AND DataProgress AND PauseResponder AND it will return
+// datatransfer.ErrPause as the error
+func (c *Channels) fireProgressEvent(chid datatransfer.ChannelID, evt datatransfer.EventCode, progressEvt datatransfer.EventCode, delta uint64, index int64, unique bool, readFromOriginal readIndexFn, readProgress readProgressFn) error {
 	if err := c.checkChannelExists(chid, evt); err != nil {
-		return false, err
+		return err
 	}
 
-	isNewIndex, err := c.blockIndexCache.updateIfGreater(evt, chid, index, readFromOriginal)
+	pause, progress, err := c.checkEvents(chid, evt, delta, index, unique, readFromOriginal, readProgress)
+
 	if err != nil {
-		return false, err
+		return err
 	}
 
-	// If the block has not been seen before, fire the progress event
-	if unique && isNewIndex {
+	// Fire the progress event if there is progress
+	if progress {
 		if err := c.stateMachines.Send(chid, progressEvt, delta); err != nil {
-			return false, err
+			return err
 		}
 	}
 
 	// Fire the regular event
-	return unique && isNewIndex, c.stateMachines.Send(chid, evt, index)
+	if err := c.stateMachines.Send(chid, evt, index); err != nil {
+		return err
+	}
+
+	// fire the pause event if we past our data limit
+	if pause {
+		// pause. Data limits only exist on the responder, so we always pause the responder
+		if err := c.stateMachines.Send(chid, datatransfer.PauseResponder); err != nil {
+			return err
+		}
+		// return a pause error so the transfer knows to pause
+		return datatransfer.ErrPause
+	}
+	return nil
+}
+
+func (c *Channels) checkEvents(chid datatransfer.ChannelID, evt datatransfer.EventCode, delta uint64, index int64, unique bool, readFromOriginal readIndexFn, readProgress readProgressFn) (pause bool, progress bool, err error) {
+
+	// if this is not a unique block, no data progress is made, return
+	if !unique {
+		return
+	}
+
+	// check if data progress is made
+	progress, err = c.blockIndexCache.updateIfGreater(evt, chid, index, readFromOriginal)
+	if err != nil {
+		return false, false, err
+	}
+
+	// if no data progress, return
+	if !progress {
+		return
+	}
+
+	// if we don't check data limits on this function, return
+	if readProgress == nil {
```

Comment:
what are the cases where this is `nil`? is it just `DataSent` calls? and if so, the only value in calling this function is to set the progress of the sent index, correct?

very minor nit - the naming of this function could be improved, current name doesn't imply mutation but that's one of its two core functions, `updateIndexAndCheckEvents()` might be the verbose option.

Reply:
yea agree on all fronts. I'm inclined to wait and update when the transport refactor happens, cause I think data limit logic will move downward to the transport